core/src/main/java/org/pentaho/di/core/row/value/ValueMetaTimestamp.java

```
@@ -0,0 +1,709 @@
+    }
+
+    switch ( storageType ) {
+      case STORAGE_TYPE_NORMAL:
+        return convertStringToBinaryString( getString( object ) );
+      case STORAGE_TYPE_BINARY_STRING:
+        return convertStringToBinaryString( (String) convertBinaryStringToNativeType( (byte[]) object ) );
+      case STORAGE_TYPE_INDEXED:
+        return convertStringToBinaryString( getString( index[ ( (Integer) object ).intValue() ] ) );
+      default:
+        throw new KettleValueException( toString() + " : Unknown storage type " + storageType + " specified." );
+    }
+  }
+
+  @Override
+  public void writeData( DataOutputStream outputStream, Object object ) throws KettleFileException {
+    try {
+      // Is the value NULL?
+      outputStream.writeBoolean( object == null );
+
+      if ( object != null ) {
+        switch ( storageType ) {
+          case STORAGE_TYPE_NORMAL:
+            // Handle Content -- only when not NULL
+            Timestamp timestamp = convertDateToTimestamp( (Date) object );
+
+            outputStream.writeLong( timestamp.getTime() );
+            outputStream.writeInt( timestamp.getNanos() );
+            break;
+
+          case STORAGE_TYPE_BINARY_STRING:
+            // Handle binary string content -- only when not NULL
+            // In this case, we opt not to convert anything at all for speed.
+            // That way, we can save on CPU power.
+            // Since the streams can be compressed, volume shouldn't be an issue
+            // at all.
+            //
+            writeBinaryString( outputStream, (byte[]) object );
+            break;
+
+          case STORAGE_TYPE_INDEXED:
+            writeInteger( outputStream, (Integer) object ); // just an index
+            break;
+
+          default:
+            throw new KettleFileException( toString() + " : Unknown storage type " + getStorageType() );
+        }
+      }
+    } catch ( ClassCastException e ) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 617.